server/rpc.go

```diff
@@ -0,0 +1,462 @@
+	if err != nil {
+		return nil, err
+	}
+	sids, err := orch.StreamIDs(job)
+	if err != nil {
+		return nil, err
+	}
+	stringStreamIds := make(map[string]string)
+	for i, s := range sids {
+		stringStreamIds[s.String()] = job.Profiles[i].Name
+	}
+
+	tr := TranscoderInfo{
+		Transcoder:  orch.Transcoder(),
+		AuthType:    AuthType_LPE,
+		Credentials: creds,
+		StreamIds:   stringStreamIds,
+	}
+	return &tr, nil
+}
+
+func (orch *orchestrator) ServeSegment(w http.ResponseWriter, r *http.Request) {
+	// check the credentials from the orchestrator
+	authType := r.Header.Get("Authorization")
+	creds := r.Header.Get("Credentials")
+	if AuthType_LPE != authType {
+		glog.Error("Invalid auth type ", authType)
+		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		return
+	}
+	job, err := verifyToken(orch, creds)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusForbidden)
+		return
+	}
+
+	// check the segment sig from the broadcaster
+	seg := r.Header.Get("Livepeer-Segment")
+	segData, err := verifySegCreds(job, seg)
+	if err != nil {
+		glog.Error("Could not verify segment creds")
+		http.Error(w, err.Error(), http.StatusForbidden)
+		return
+	}
+
+	// download the segment and check the hash
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		glog.Error("Could not read request body")
+		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		return
+	}
+	hash := crypto.Keccak256(data)
+	if !bytes.Equal(hash, segData.Hash) {
+		glog.Error("Mismatched hash for body; rejecting")
+		http.Error(w, "Forbidden", http.StatusForbidden)
+		return
+	}
+
+	ss := core.SignedSegment{
+		Seg: stream.HLSSegment{
+			SeqNo:    uint64(segData.Seq),
+			Name:     "",
+			Data:     data,
+			Duration: 0.,
+		},
+		Sig: segData.Sig,
+	}
+	if err := orch.TranscodeSeg(job, &ss); err != nil {
+		glog.Error("Could not transcode ", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write([]byte("The segment has been successfully transcoded."))
+}
+
+type lphttp struct {
+	orchestrator *grpc.Server
+	transcoder   *http.ServeMux
+}
+
+func (h *lphttp) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ct := r.Header.Get("Content-Type")
+	if r.ProtoMajor == 2 && strings.HasPrefix(ct, "application/grpc") {
+		h.orchestrator.ServeHTTP(w, r)
+	} else {
+		h.transcoder.ServeHTTP(w, r)
+	}
+}
+
+func StartTranscodeServer(bind string, publicURI *url.URL, node *core.LivepeerNode) {
+	s := grpc.NewServer()
+	addr := node.Eth.Account().Address
+	orch := orchestrator{transcoder: publicURI.String(), node: node, address: addr}
+	RegisterOrchestratorServer(s, &orch)
+	lp := lphttp{
+		orchestrator: s,
+		transcoder:   http.NewServeMux(),
+	}
+	lp.transcoder.HandleFunc("/segment", orch.ServeSegment)
+
+	cert, key, err := getCert(publicURI, node.WorkDir)
+	if err != nil {
+		return // XXX return error
+	}
+
+	glog.Info("Listening for RPC on ", bind)
+	http.ListenAndServeTLS(bind, cert, key, &lp)
```

Comment:
I don't think the `http` package level convenience functions like `ListenAndServeTLS` by default enforce a read/write timeouts from clients - perhaps we should consider configuring default timeout values for the HTTP server to drop slow clients if they take forever to read/write?

Reply:
Sure, set it to 4 seconds for now, for the client and server code. If a RTT takes longer than that, segments are guaranteed to be backed up.